components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java

```diff
@@ -0,0 +1,160 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.config.model.graph;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Supervises the execution of any script engine, and kills the thread if the time taken is too much.
+ */
+public class JSExecutionSupervisor {
+
+    private static final Log LOG = LogFactory.getLog(JSExecutionSupervisor.class);
+    private static final String JS_EXECUTION_MONITOR = "JS-Exec-Monitor";
+    private final long TIMEOUT_IN_MILLIS;
+    private final int THREAD_COUNT;
+    private long taskExecutionRateInMillis = 100L;
+    private Map<String, ScheduledFuture> currentScriptExecutions = new HashMap<>();
+    private Map<String, MonitoringTask> monitoredTasks = new HashMap<>();
+    private ScheduledExecutorService monitoringService;
+
+    public JSExecutionSupervisor(int threadCount, long timeoutInMillis) {
+
+        this.THREAD_COUNT = threadCount;
+        this.TIMEOUT_IN_MILLIS = timeoutInMillis;
+
+        if (taskExecutionRateInMillis > timeoutInMillis) {
+            taskExecutionRateInMillis = timeoutInMillis;
+        }
+
+        init();
+    }
+
+    private void init() {
+
+        monitoringService = new ScheduledThreadPoolExecutor(THREAD_COUNT, r -> new Thread(r, JS_EXECUTION_MONITOR));
```

Comment:
do we really need a separate init method?

Reply:
updated